### DISCUSSION TITLE
service-fn blocks main thread even when running inside another thread

### DISCUSSION THREAD

**Discussion by J0sueTM** [moclojer/moclojer#270] [Q&A] [answered]
I've been trying to setup some tests that automatically stop the server when finishing by running (helpers/service-fn) in a new thread or `go` block:

```clojure
  (deftest my-test
    (let [server (atom nil)]

      (async/go
        (reset! server (helpers/service-fn "test/com/moclojer/resources/moclojer-v2.yaml")))

      ;; ... do stuff

      (p.http/stop @server)))
```

However, that doesn't seem to work. Even if i use thread, future, or whatever. I can't get to set the server atom with the new value.

---

I'm quite sure it's a problem with the way I'm trying to do `reset!`, but I would like to ask you guys if you can see the problem. Thanks in advance.

This problem arose during the development of https://github.com/moclojer/moclojer/pull/262

**Comment by avelino** [comment]
@J0sueTM wouldn't that be what you're trying to do?

https://github.com/moclojer/moclojer/blob/660da616f274e179517b7ab19b1898ac308fbb97/test/com/moclojer/framework_test.clj#L8-L23

if not, I think I didn't understand the problem you are reporting

**Reply by J0sueTM** [reply]
That works fine, but not for my REPL test workflow. `(service-fn)` opens a webserver, which's process becomes a child of the current REPL process. So, in order for me to run the next test, I need to close that open webserver, using `(pedestal.http/stop)`.

You probably run the tests from the terminal with `clj -M:test`, right? The clj process automatically kills the webserver process, so it might have not been a problem until now for you.

**Reply by avelino** [reply]
I understand, it really makes sense to have `start` and `stop` to encourage the use of **REPL**
I don't see this as a **bug** or a lack of **feature implementation** lol

**Reply by J0sueTM** [reply]
Oh, definitely. sorry for not explaining my intentions. I created this discussion because this problem can be the same of someone else in the future. Since it's not a bug, it didn't make sense as well to open an issue. Anyways, my bad for tagging this as a bug :sweat_smile:

**Reply by J0sueTM** [reply, marked as answer]
@avelino After fiddling around with service-fn, I discovered that we actually don't need to create a webserver in order to use moclojer. All we need to do is "start" it as a pedestal interceptor:

```clojure
(-> (helpers/service-fn
     "test/com/moclojer/resources/moclojer-v2.yaml"
     {:start? false :join? false})
    (response-for :get "/users/avelino/2222")
    ...
```
